docs/language/informal/covariant-overrides.md

```diff
@@ -0,0 +1,742 @@
+some type checks are not guaranteed to succeed based on static checks, so
+they must be performed at run time.
+
+In particular, when a method with a covariant parameter is invoked, it is not
+possible to guarantee based on the static type check at call sites that
+the actual argument will have the type which is declared for that parameter
+in the method that is actually invoked. A dynamic type check must then be
+performed to verify that the actual argument is null or has that type,
+and a `CastError` must be thrown if the check fails.
+
+In other words, the static checks enforce the original type, and the dynamic
+check enforces the desired type, and both checks are required because the
+desired type can be a proper subtype of the original type. Here is an example:
+
+```dart
+Widget widget = new RadioGroup();
+try {
+  widget.addChild(new Widget());
+} on CastError {
+  print("Caught!"); // Gets printed.
+}
+```
+
+In practice, a compiler may generate code such that every covariant parameter
+in each method implementation gets checked before the body of the method is
+executed. This is a correct implementation strategy because the dynamic check
+is performed with the desired type and it is performed for all invocations,
+no matter which original type the call site had.
+
+As an optimization, the dynamic check could be omitted in some cases, because
+it is guaranteed to succeed. For instance, we may know statically that the
+original and the desired type are the same, because the receiver's type is
+known exactly:
+
+```dart
+new RadioGroup().add(myRadioButton())
+```
+
+#### The runtime type of a tear-off
+
+For any given declaration *D* of a method `m`, the reified type
+obtained by tearing off `m` from a receiver whose dynamic type is a
+class `C` wherein *D* is the most specific declaration of `m` (that
+is, *D* has not been overridden by any other method declaration in
+`C`) can be computed as follows: Collect the set of all declarations
+of `m` in direct and indirect supertypes of `C`. For the reified
+return type, choose the greatest lower bound of all declared return
+types; for the reified parameter type of each covariant parameter,
+choose the least upper bound of all declared types for that parameter
+(including any declarations of that parameter which are not
+covariant).
+
+As a consequence, the reified type of a torn off method is at least as
+specific as every possible statically known type for that method, and
+that means that it is safe to assign the torn off method to a
+variable whose declared type is the statically known type of that
+method, no matter which supertype of the dynamic type of the receiver
+is the statically known receiver type.
+
+(Note that the computation of a "least upper bound" in Dart does not
+in fact yield a *least* upper bound, but it does yield some *upper
+bound*, which is sufficient to ensure that this safety property holds.)
```

Comment:
The LUB/MAX we are discussing elsewhere has the property that it sometimes has to give up. In that case, we should probably use `Object` here (which is always an upper bound). For inference, we want something else that defaulting to `Object`, so it's not part of the "LUB" computation.